packages/components/src/internal/util/RouteLeave.tsx

```diff
@@ -1,96 +1,135 @@
-import React from 'react';
-import { Location } from 'history';
-import { WithRouterProps } from 'react-router';
+import React, { ComponentType, FC, useCallback, useEffect, useMemo, useRef } from 'react';
+import { InjectedRouter, withRouter, WithRouterProps, PlainRoute } from 'react-router';
 
-import { AppURL } from '../..';
-
-import { BeforeUnload } from './BeforeUnload';
-
-export const ON_LEAVE_DIRTY_STATE_MESSAGE =
-    'You have unsaved changes that will be lost. Are you sure you want to continue?';
+export const CONFIRM_MESSAGE = 'You have unsaved changes that will be lost. Are you sure you want to continue?';
 
 /**
- * This function can be used as the callback for react-router's setRouteLeaveHook.  It should be preferred
- * over a callback that simply returns the confirm message because with react-router v3.x, the URL route
- * will have already been changed by the time this confirm is shown and will not be reset if the user does not confirm.
- * If the user tries to click on the initial link again after canceling, nothing will happen because the URL
- * in the browser will not change. (Issue 39633).
- * See also https://stackoverflow.com/questions/32841757/detecting-user-leaving-page-with-react-router
- *
- * TODO: Seems like there are some additional tools in newer versions of react-router:
- *  https://stackoverflow.com/questions/62792342/in-react-router-v6-how-to-check-form-is-dirty-before-leaving-page-route
- *
- * @param currentLocation the location of the current page
- * @param event object triggering leave transition
+ * @deprecated: use useRouteLeave or withRouteLeave instead
+ * This function can be used as the callback for react-router's setRouteLeaveHook. See notes in useRouteLeave re: Issue
+ * 42101 in order to understand the caveats that come with using this function.
  */
-export function confirmLeaveWhenDirty(currentLocation: Location, event?: any): boolean {
-    const result = confirm(ON_LEAVE_DIRTY_STATE_MESSAGE);
-
-    // navigation canceled, pushing the previous path
-    if (!result && currentLocation) {
-        const appURL = AppURL.create(...currentLocation.pathname.substring(1).split('/').map(decodeURIComponent));
-
-        // Issue 41419: handle different cases for route leave PUSH vs POP events
-        if (event?.action !== 'POP') {
-            window.history.replaceState(null, null, appURL.toHref() + currentLocation.search);
-        } else {
-            window.history.back();
-        }
+export function confirmLeaveWhenDirty(): boolean {
```

Comment:
Is this actually still in use anywhere?

Reply:
I think so, but let me double check. I may have been confusing it with BeforeUnload

Reply:
Looks like it is no longer used, BeforeUnload is though. I will remove this and update release notes